### DISCUSSION TITLE
How do I model conditional constraints in YALMIP？

### DISCUSSION THREAD

**Discussion by YANGseuJW** [yalmip/YALMIP#1435] [Problems using YALMIP] [answered]
Dear sir！
I encountered some issues while describing conditional constraint.
Decision variables are as follows:
x=sdpvar(1,1);y=sdpvar(1,1);z=sdpvar(1,1);w1=sdpvar(1,1);w2=sdpvar(1,1);
u=binvar(1,1);k=sdpvar(1,1);
Constraints are as follows:
if x>y, z=w1, else z>=w2.
if u=0, k=0
if k=0, u=0
where w1,w2
I use this method to establish such constraint (part):
implies(x>=y+eps,z==w1);
implies(x<=y,z>w2);
implies(u==0,k==0);
implies(k==0,u==0);
There is no problem using yalmip to build the model, but when calling gurobi, although it can be solved, there are significant solutions that violate this constraint.May I ask what caused this?

**Comment by YANGseuJW** [comment]
correct: implies(x<=y,z>=w2);

**Comment by johanlofberg** [comment]
Binary implications should simply be written as `implies(1-u, k == 0)`

Disjoints constraints should always explicitly be made disjoint, as solvers easily can see two disjoint constraints to be active/inactive due to numerical tolerances, i.e. 

`implies(case1,[x>=y, z == w1]) + implies(case2,[x<=y, z>=w2]) + [case1+case2 == 1]`

for two binary variables case1 and case2

https://yalmip.github.io/tutorial/logicprogramming

**Reply by YANGseuJW** [reply]
Thank you, sir, I'll give it a try.

**Reply by YANGseuJW** [reply]
Sir, How to obtain binary variables case1 and case2?

**Reply by johanlofberg** [reply]
Just like you define u

**Reply by YANGseuJW** [reply]
I know binvar, but I don't know the specific meaning of case1 and case2. I think z==w1 is determined by the case of x>=y+eps (eps is a very small positive number used to represent strict constraints of x>y), and z>=w2 is determined by the case of x<=y. They should not be parallel constraints determined by case1 or case2.

**Reply by johanlofberg** [reply, marked as answer]
You define case1 and case2 as binary variables, and as only one of them can be true, the two disjoint cases you list cannot occur at the same time but one has to occur (and of course you will have to add an eps in some direction if you have that as you cannot work with strict cases)

**Reply by johanlofberg** [reply]
https://yalmip.github.io/example/modellingif/

**Reply by YANGseuJW** [reply]
I understand, thank you, sir
